src/history/file_backed.rs

```diff
@@ -0,0 +1,543 @@
+    }
+}
+
+impl History for FileBackedHistory {}
+
+impl HistoryAppender for FileBackedHistory {
+    /// Appends an entry if non-empty and not repetition of the previous entry.
+    /// Resets the browsing cursor to the default state in front of the most recent entry.
+    ///
+    fn append(&mut self, entry: String) {
+        // Don't append if the preceding value is identical or the string empty
+        if self
+            .entries
+            .back()
+            .map_or(true, |previous| previous != &entry)
+            && !entry.is_empty()
+        {
+            if self.entries.len() == self.capacity {
+                // History is "full", so we delete the oldest entry first,
+                // before adding a new one.
+                self.entries.pop_front();
+                self.len_on_disk = self.len_on_disk.saturating_sub(1);
+                self.truncate_file = true;
+            }
+            self.entries.push_back(entry);
+        }
+        self.reset_cursor()
+    }
+
+    fn iter_chronologic(&self) -> Iter<'_, String> {
+        self.entries.iter()
+    }
+}
+
+impl HistoryView for FileBackedHistory {
+    fn back(&mut self) {
+        match self.query.clone() {
+            HistoryNavigationQuery::Normal => {
+                if self.cursor > 0 {
+                    self.cursor -= 1;
+                }
+            }
+            HistoryNavigationQuery::PrefixSearch(prefix) => {
+                let mut cursor = self.cursor;
+
+                let previous_match = self.string_at_cursor();
+
+                while cursor > 0 {
+                    cursor -= 1;
+                    let entry = &self.entries[cursor];
+                    if entry.starts_with(&prefix) {
+                        if previous_match
+                            // TODO Get rid of this clone
+                            .clone()
+                            .map_or(false, |value| &value == entry)
+                        {
+                            continue;
+                        } else {
+                            break;
+                        }
+                    }
+                }
+
+                self.cursor = cursor;
+            }
+            HistoryNavigationQuery::SubstringSearch(substring) => {
+                let mut cursor = self.cursor;
+                let previous_match = self.string_at_cursor();
+
+                while cursor > 0 {
+                    cursor -= 1;
+                    let entry = &self.entries[cursor];
+                    if entry.contains(&substring) {
+                        if previous_match
+                            // TODO Get rid of this clone
+                            .clone()
+                            .map_or(false, |value| &value == entry)
+                        {
+                            continue;
+                        } else {
+                            break;
+                        }
+                    }
+                }
```

Comment:
Lot's of duplication. Any way to avoid this?

Reply:
If they're identical you can carry two difference cases in a match like `A | B => { ... }` Is that what you're asking?  You'd need to bind to the same name, but assuming you do and they're the same type, the rest would become a single match body

Reply:
Oh I see where they differ now. Unfortunately, the only way I know of is to introduce macros here or functions that take functions as parameters. That could work, but it's difficult to do that but also keep the code readable.

Reply:
I would also like to avoid the usage of macros. Maybe I can pass in matching criteria as a function and then have a generic function perform all the other operations. Though this is something that can be picked up in a later PR.

Reply:
https://github.com/jonathandturner/reedline/pull/92/commits/0cc04453f76bc82d74b2ee26bcb20e454559433e